demo_fastapi/sales.py

```diff
@@ -1,83 +1,168 @@
+            detail="Please enter buyer's name "
+            "(max 50 characters, without numbers in the name).",
         )
-    if not r.match(datestamp):
+
+    # if everything ok, make an order
+    order_id = random.randrange(10**8)
+    order.id = order_id
+
+    # save to the database
+    command = "INSERT INTO ShoppingList VALUES (?,?,?,?,?)"
+    odbc_execute_command(
+        cnxn,
+        command,
+        order.datestamp,
+        order.buyer,
+        order.apples,
+        order.oranges,
+        order.id,
+    )
+
+    return order
+
+
+@app.put("/orders/")
+def update_order(order: Order):
+    """update order with ID = order_id"""
+    # establish connection to the database
+    cnxn = app.state.connection_manager.connection
+
+    # check what were the old values for the order with id = order.id
+    command = "SELECT * FROM ShoppingList WHERE id = ?"
+    cursor = odbc_execute_command(cnxn, command, order.id)
+    row = cursor.fetchone()
+    if row is None:
         raise HTTPException(
             status_code=422,
-            detail="Date in wrong format! It should be yyyy/mm/dd (no spaces).",
+            detail="Provided order (id) is not in the database table ShoppingList. "
+            "Please provide another id!",
         )
 
-    try:
-        date_ = datetime.strptime(datestamp, "%Y/%m/%d")
-    except ValueError as err:
-        raise HTTPException(status_code=422, detail=str(err))
+    # update with new value from input variable order, or keep the previous value
+    datestamp = order.datestamp or row.datestamp
+    buyer = order.buyer or row.buyer
+    apples = order.apples or row.apples
+    oranges = order.oranges or row.oranges
 
-    if date_ < datetime(2000, 1, 1):
+    # update working table
+    command = (
+        "UPDATE ShoppingList SET datestamp = ?, buyer = ?, apples = ?, oranges "
+        "= ? WHERE id = ?"
+    )
+    odbc_execute_command(cnxn, command, datestamp, buyer, apples, oranges, order.id)
+
+    # if the change passed, log the last version in the log table
+    date_changed = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+
+    command = "INSERT INTO ShoppingListLog VALUES (?,?,?,?,?,?)"
+    odbc_execute_command(
+        cnxn,
+        command,
+        row.datestamp,
+        row.buyer,
+        row.apples,
+        row.oranges,
+        row.id,
+        date_changed,
+    )
+
+    return Order(
+        id=order.id, datestamp=datestamp, buyer=buyer, apples=apples, oranges=oranges
+    )
+
+
+@app.get("/orders/")
+def read_order(order_id: int):
+    """Read the order with ID = order_id from the database and print it in the app"""
+    # establish connection
+    cnxn = app.state.connection_manager.connection
+
+    # read the order values from the database table
+    cursor = odbc_execute_command(
+        cnxn, "SELECT * FROM ShoppingList WHERE id = ?", order_id
+    )
+    row = cursor.fetchone()
+    if row is None:
         raise HTTPException(
             status_code=422,
```

Comment:
status code: 422 = “unprocessable entity" which is when what is given is of wrong format, but here the problem is diff so maybe we should change this?

Reply:
I looked into the [code table](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) and wasn't sure what exception to use, other than 422 or 404 (resource not found). What is the best to use here?

Reply:
404 is better :-)